Lesson Number 4 (#44)
dwyl/learn-flutter: Flutter_TDD_Architecture_Course_notes.md

And now the **second** test will pass.

![Second_test_pass](https://user-images.githubusercontent.com/27420533/74590474-aa92e180-5006-11ea-8a88-ef8a9a6311e7.png)

#### 4. Data Layer Overview & Models
> Video 4: https://www.youtube.com/watch?v=keaTZ9M_U1A&t=1516s

> In the `fourth` tutorial start by giving a brief explanation of how the models work.

Reviewer: ```suggestion
The `fourth` tutorial starts by giving a brief explanation of how **models** work.
```

> The models are like entities but with more features in the case of this specific application will deal with making the data conversation in Json to Dart.

Reviewer: ```suggestion
Models are like entities but with more features. In the case of this application our model will have a function to convert `JSON` data to a Dart Map.
```

> After this explanation is necessary to create a file inside the **models** folder with the name `number_trivia_model.dart`.

Reviewer: The words "After this explanation" do not add to the remainder of the sentence. 
You can convey the same meaning/message with fewer words:
```suggestion
Create a file in the **models** folder named `number_trivia_model.dart`.
```

> And then its test version inside the **test** folder with the name `number_trivia_model_test.dart`.

Reviewer: ```suggestion
Create the corresponding test file in the **test** folder with the name `number_trivia_model_test.dart`.
Here you can see the naming convention follows a predictable pattern which makes locating files in a project very fast.
```

> This will be the code that will be inside the test file.

Reviewer: ```suggestion
Add the following code to the `number_trivia_model_test.dart` file:
```

```dart
import 'package:clean_architecture_tdd_prep/features/number_trivia/data/models/number_trivia_model.dart';
import 'package:clean_architecture_tdd_prep/features/number_trivia/domain/entities/number_trivia.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  final tNumberTriviaModel = NumberTriviaModel(number: 1, text: 'Test Text');

  test(
    'should be a subclass of NumberTrivia entity',
    () async {
      // assert
      expect(tNumberTriviaModel, isA<NumberTrivia>());
    },
  );
}
```

> The test will **not pass** so we will go to the `number_trivia_model.dart` file and pass some parameters.

Reviewer: ```suggestion
The test will ***not pass***. This is _expected_ because we have not yet implemented the `NumberTriviaModel` class.
Open the `number_trivia_model.dart` file and add the following code:
```

```dart
import 'package:meta/meta.dart';

import '../../domain/entities/number_trivia.dart';

class NumberTriviaModel extends NumberTrivia {
  NumberTriviaModel({
    @required String text,
    @required int number,
  }) : super(
          text: text,
          number: number,
        );
}
```

> Then it is necessary to create inside the test folder a folder called `fixtures` that will contain two files one `trivia.json` and the other `trivia_double.json`.

Reviewer: What you had written here was _good_ but _incomplete_. 
We want to give more instructions and information to the learner/reader:
```suggestion
Once you have added this code to the `number_trivia_model.dart` file, re-run the tests: `flutter test`

Given that our Model will transform `JSON` data, we need some some test data.
Test data is commonly referred to as [**fixtures**](https://en.wikipedia.org/wiki/Test_fixture#Software).
In the test folder, create a new folder called **`fixtures`** that will contain two files 
one called `trivia.json` and the other `trivia_double.json`.
```

![fixtures](https://user-images.githubusercontent.com/27420533/75034654-94bd6a80-54a5-11ea-96a2-7c7bd2942f8a.png)

This will be the code for `trivia.json`.

```dart
{
  "text": "Test Text",
  "number": 1,
  "found": true,
  "type": "trivia"
}
```

> And for `trivia_double.json`.

Reviewer: ```suggestion
And for `trivia_double.json` the only difference is that the **"number"** value is a [double](https://dart.dev/articles/archive/numeric-computation) `1.0` instead of an integer `1`:
```

Open the `number_trivia_model.dart` file and add the following code:

```dart
import 'package:meta/meta.dart';

import '../../domain/entities/number_trivia.dart';

class NumberTriviaModel extends NumberTrivia {
  NumberTriviaModel({
    @required String text,
    @required int number,
  }) : super(
          text: text,
          number: number,
        );
}
```
Once you have added this code to the `number_trivia_model.dart` file, re-run the tests: `flutter test`

Given that our Model will transform `JSON` data, we need some some test data.
Test data is commonly referred to as [**fixtures**](https://en.wikipedia.org/wiki/Test_fixture#Software).
In the test folder, create a new folder called **`fixtures`** that will contain two files 
one called `trivia.json` and the other `trivia_double.json`.

![fixtures](https://user-images.githubusercontent.com/27420533/75034654-94bd6a80-54a5-11ea-96a2-7c7bd2942f8a.png)

This will be the code for `trivia.json`.

```dart
{
  "text": "Test Text",
  "number": 1,
  "found": true,
  "type": "trivia"
}
```
And for `trivia_double.json`.

```dart
{
  "text": "Test Text",
  "number": 1.0,
  "found": true,
  "type": "trivia"
}
```

> Now we have to create a file that will allow us to turn the json's answers into integer or double for String.

Reviewer: ```suggestion
Next we need to create a file that will turn the `JSON` answers into `integer` or `double`.
```

> Then we create a file called `fixture_reader.dart`.

Reviewer: ```suggestion
Create a file called `fixture_reader.dart` and add the following code:
```

> The code for `fixture_reader.dart`.

Reviewer: ```suggestion

```

```dart
class NumberTriviaModel extends NumberTrivia {
  ...
  Map<String, dynamic> toJson() {
    return {
      'text': text,
      'number': number,
    };
  }
}
```
Now all the tests **pass**.

#### 5. Contracts of Data Sources
> Video 5: https://www.youtube.com/watch?v=m_lkZo6CYcs&t=1s

In the `fifth` tutorial an explanation is given about the importance of `contracts`.
The contracts will allow the files to communicate directly with the **repository**.

Inside the `repositories` folder create `number_trivia_repository_impl.dart`.
This is the code for the file.

```dart
import 'package:dartz/dartz.dart';

import '../../../../core/error/failure.dart';
import '../../domain/entities/number_trivia.dart';
import '../../domain/repositories/number_trivia_repository.dart';

class NumberTriviaRepositoryImpl implements NumberTriviaRepository {
  @override
  Future<Either<Failure, NumberTrivia>> getConcreteNumberTrivia(int number) {
    // TODO: implement getConcreteNumberTrivia
    return null;
  }

  @override
  Future<Either<Failure, NumberTrivia>> getRandomNumberTrivia() {
    // TODO: implement getRandomNumberTrivia
    return null;
  }
}
```

> Inside the `platform` folder create a file called `network_info.dart`with the following code.

Reviewer: ```suggestion
Inside the `platform` folder create a file called `network_info.dart` with the following code:
```

```dart
abstract class NetworkInfo {
  Future<bool> get isConnected;
}
```
#### Remote Data Source

> In this class the value returned will be different from the others and will return only a `NumberTriviaModel` value and the errors will be treated as `Exceptions`.

Reviewer: This sentence is 27 words. It has two instances of the word "**and**". That is usually a sign that you are combining too many thoughts into the same sentence. It's much better for the reader to split it into two or cut back the one large sentence.
e.g:
```suggestion
In this class the returned value will be a `NumberTriviaModel` value and errors will be treated as `Exceptions`.
```

> Then we create the file `number_trivia_remote_data_source.dart`,with the following code.

Reviewer: ```suggestion
Create the file `number_trivia_remote_data_source.dart` with the following code:
```

```dart
import '../models/number_trivia_model.dart';

abstract class NumberTriviaRemoteDataSource {
  /// Calls the http://numbersapi.com/{number} endpoint.
  ///
  /// Throws a [ServerException] for all error codes.
  Future<NumberTriviaModel> getConcreteNumberTrivia(int number);

  /// Calls the http://numbersapi.com/random endpoint.
  ///
  /// Throws a [ServerException] for all error codes.
  Future<NumberTriviaModel> getRandomNumberTrivia();
}

```
#### Exceptions And Failures

> Whenever we don't have a number to show it will be a `ServerException`, then we should create it to be used.

Reviewer: ```suggestion
Whenever we don't have a number to show it will be a `ServerException`. We need a class to handle these exceptions.
```

> Inside the error folder create the file `exception.dart`.

Reviewer: ```suggestion
In the **`/error`** folder, create the file `exception.dart` and add the following code:
```

> Code for `Exception.dart`.

Reviewer: don't capitalise the file name if it's not capitalised in the proceeding line. 
```suggestion

```

```dart
class ServerException implements Exception {}

class CacheException implements Exception {}
```

> This is the code for the `failure.dart` file that also has to be created.

Reviewer: _Where_ does the `failure.dart` file need to be created? Don't skip information. 💭

abstract class Failure extends Equatable {
  Failure([List properties = const <dynamic>[]]) : super(properties);
}

// General failures
class ServerFailure extends Failure {}

class CacheFailure extends Failure {}
```
Now we need to create a file for the `NumberTriviaRepositoryImpl` tests.
This file will be named `number_trivia_repository_impl_test.dart`.
Then we prepare the file for the test.

```dart
class MockRemoteDataSource extends Mock
    implements NumberTriviaRemoteDataSource {}

class MockLocalDataSource extends Mock implements NumberTriviaLocalDataSource {}

class MockNetworkInfo extends Mock implements NetworkInfo {}

void main() {
  NumberTriviaRepositoryImpl repository;
  MockRemoteDataSource mockRemoteDataSource;
  MockLocalDataSource mockLocalDataSource;
  MockNetworkInfo mockNetworkInfo;

  setUp(() {
    mockRemoteDataSource = MockRemoteDataSource();
    mockLocalDataSource = MockLocalDataSource();
    mockNetworkInfo = MockNetworkInfo();
    repository = NumberTriviaRepositoryImpl(
      remoteDataSource: mockRemoteDataSource,
      localDataSource: mockLocalDataSource,
      networkInfo: mockNetworkInfo,
    );
  });
}
```

> If we run the program now it will give error because we have to pass all parameters also to the file `number_trivia_repository_impl.dart`.

Reviewer: ```suggestion
If we run the test now it will give an error because have not yet written the implementation. 
Open the `number_trivia_repository_impl.dart` file and add the following code:
```

```dart
import 'package:dartz/dartz.dart';
import 'package:meta/meta.dart';

import '../../../../core/error/failure.dart';
import '../../../../core/platform/network_info.dart';
import '../../domain/entities/number_trivia.dart';
import '../../domain/repositories/number_trivia_repository.dart';
import '../datasources/number_trivia_local_data_source.dart';
import '../datasources/number_trivia_remote_data_source.dart';

class NumberTriviaRepositoryImpl implements NumberTriviaRepository {
  final NumberTriviaRemoteDataSource remoteDataSource;
  final NumberTriviaLocalDataSource localDataSource;
  final NetworkInfo networkInfo;

  NumberTriviaRepositoryImpl({
    @required this.remoteDataSource,
    @required this.localDataSource,
    @required this.networkInfo,
  });

  @override
  Future<Either<Failure, NumberTrivia>> getConcreteNumberTrivia(int number) {
    // TODO: implement getConcreteNumberTrivia
    return null;
  }

  @override
  Future<Either<Failure, NumberTrivia>> getRandomNumberTrivia() {
    // TODO: implement getRandomNumberTrivia
    return null;
  }
}
```
#### 6. Repository Implementation
> Video 6: https://www.youtube.com/watch?v=bfEKPKKy9dA

> In the `sixth` tutorial start by explaining that we will have to know if the user has an internet connection and for that a test will be done.

Reviewer: Grammar and excess word count. 
![image](https://user-images.githubusercontent.com/194400/75149177-4f917680-56f9-11ea-801c-28c516837481.png)

```suggestion
The `sixth` video starts by explaining that we will have to know if the user has an internet connection.
```

#### Device Is Online

```dart
group('getConcreteNumberTrivia', () {
  // DATA FOR THE MOCKS AND ASSERTIONS
  // We'll use these three variables throughout all the tests
  final tNumber = 1;
  final tNumberTriviaModel =
      NumberTriviaModel(number: tNumber, text: 'test trivia');
  final NumberTrivia tNumberTrivia = tNumberTriviaModel;

  test('should check if the device is online', () {
    //arrange
    when(mockNetworkInfo.isConnected).thenAnswer((_) async => true);
    // act
    repository.getConcreteNumberTrivia(tNumber);
    // assert
    verify(mockNetworkInfo.isConnected);
  });
});
```
The test will `fail` because we have to add features.

> Add the code to the `number_trivia_repository_impl_test.`dart.

Reviewer: Miguel, you are _clearly_ not reading what you write ***`before`*** assigning it to me.
If you did, you would spot and _fix_ these kinds of _basic_ errors:
```suggestion
Add the code to the `number_trivia_repository_impl_test.dart` file:
```

group('device is online', () {
  // This setUp applies only to the 'device is online' group
  setUp(() {
    when(mockNetworkInfo.isConnected).thenAnswer((_) async => true);
  });

  test(
    'should return remote data when the call to remote data source is successful',
    () async {
      // arrange
      when(mockRemoteDataSource.getConcreteNumberTrivia(tNumber))
          .thenAnswer((_) async => tNumberTriviaModel);
      // act
      final result = await repository.getConcreteNumberTrivia(tNumber);
      // assert
      verify(mockRemoteDataSource.getConcreteNumberTrivia(tNumber));
      expect(result, equals(Right(tNumberTrivia)));
    },
  );
});
```
Now we need to do some tests and add some features.

```dart
test(
  'should cache the data locally when the call to remote data source is successful',
  () async {
    // arrange
    when(mockRemoteDataSource.getConcreteNumberTrivia(tNumber))
        .thenAnswer((_) async => tNumberTriviaModel);
    // act
    await repository.getConcreteNumberTrivia(tNumber);
    // assert
    verify(mockRemoteDataSource.getConcreteNumberTrivia(tNumber));
    verify(mockLocalDataSource.cacheNumberTrivia(tNumberTrivia));
  },
);
```

```dart

> impl.dart

Reviewer: `impl.dart` is a _file_ not a line of code. Why is it in this code block?

Author: @nelsonic It was a mistake when I was writing the name of the file he passed down.

        expect(result, equals(Right(tNumberTrivia)));
      },
    );

    test(
      'should return CacheFailure when there is no cached data present',
      () async {
        // arrange
        when(mockLocalDataSource.getLastNumberTrivia())
            .thenThrow(CacheException());
        // act
        final result = await repository.getRandomNumberTrivia();
        // assert
        verifyZeroInteractions(mockRemoteDataSource);
        verify(mockLocalDataSource.getLastNumberTrivia());
        expect(result, equals(Left(CacheFailure())));
      },
    );
  });
});
```
Do the same for the `corresponding` file.

```dart
@override
Future<Either<Failure, NumberTrivia>> getRandomNumberTrivia() async {
  if (await networkInfo.isConnected) {
    try {
      final remoteTrivia = await remoteDataSource.getRandomNumberTrivia();
      localDataSource.cacheNumberTrivia(remoteTrivia);
      return Right(remoteTrivia);
    } on ServerException {
      return Left(ServerFailure());
    }
  } else {
    try {
      final localTrivia = await localDataSource.getLastNumberTrivia();
      return Right(localTrivia);
    } on CacheException {
      return Left(CacheFailure());
    }
  }
}
```

> The last version of the `NumberTriviaRepositoryImpl`file will contain the following code.

Reviewer: ```suggestion
The last version of the `NumberTriviaRepositoryImpl` file will contain the following code:
```

import 'package:data_connection_checker/data_connection_checker.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mockito/mockito.dart';

class MockDataConnectionChecker extends Mock implements DataConnectionChecker {}

void main() {
  NetworkInfoImpl networkInfo;
  MockDataConnectionChecker mockDataConnectionChecker;

  setUp(() {
    mockDataConnectionChecker = MockDataConnectionChecker();
    networkInfo = NetworkInfoImpl(mockDataConnectionChecker);
  });

  group('isConnected', () {
    test(
      'should forward the call to DataConnectionChecker.hasConnection',
      () async {
        // arrange
        final tHasConnectionFuture = Future.value(true);

        when(mockDataConnectionChecker.hasConnection)
            .thenAnswer((_) => tHasConnectionFuture);
        // act
        // NOTICE: We're NOT awaiting the result
        final result = networkInfo.isConnected;
        // assert
        verify(mockDataConnectionChecker.hasConnection);
        // Utilizing Dart's default referential equality.
        // Only references to the same object are equal.
        expect(result, tHasConnectionFuture);
      },
    );
  });
}
```

> Then it is necessary to implement it in the corresponding file.

Reviewer: Don't skip steps. What is the file name for the implementation?

Author: @nelsonic I've already managed to change
